lib/course/middleware/etag.rb

```diff
@@ -0,0 +1,41 @@
+module Course
+  module Middleware
+    class Etag
+      attr_reader :app
+
+      def initialize(app)
+        @app = app
+      end
+
+      def call(env)
+        status, headers, body = app.call(env)
+
+        if success_status?(status) && cache?(headers)
+          digest = digest_body(body)
+          return [Statuses::NOT_MODIFIED, headers, [""]] if headers["etag"] == digest
+
+          headers["etag"] = digest
+          return [status, headers, body]
+        end
+
+        [status, headers, body]
+      end
+
+      private
+
+      def success_status?(status)
+        status == Statuses::SUCSSESS
+      end
+
+      def cache?(headers)
+        headers.include? "cache-control"
+      end
+
+      # body is an array
+      def digest_body(body)
+        body.filter { |item| !item.empty? }
```

Comment:
why not just hash the whole body?

Reply:
i did not understand the question, param for Digest::MD5.hexdigest  have to string or question about filter? i though no point for cache nil or empty string

Reply:
Though there is no harm in caching nil, and md5(body) would be easier to write, read and understand. Though you can filter much prittier: body.compact